[IO-764] IOUtils.write() throws OutOfMemoryError/NegativeArraySizeException while writing big strings

DaGeRe opened the pull request:
If Strings with a certain size are written to a stream, this fails with an exception on OpenJDK 11. 

```
java.lang.NegativeArraySizeException: -1283060862
at java.base/java.lang.StringCoding.encodeUTF8(StringCoding.java:904)
at java.base/java.lang.StringCoding.encode(StringCoding.java:449)
at java.base/java.lang.String.getBytes(String.java:964)
at org.apache.commons.io.IOUtils.write(IOUtils.java:3251)
at org.apache.commons.io.FileUtils.writeStringToFile(FileUtils.java:3541)
at org.apache.commons.io.FileUtils.writeStringToFile(FileUtils.java:3524)
```

This PR fixes the problem by using a `ByteBuffer` instead `getBytes`. 

Unfortunately, this increases test time significantly. One workaround would be to disable the test and only enable it if needed - but testing whether big Strings can be written is not possible without handling such a big string.

DaGeRe commented:
Unfortunately, this does not work on GH Actions since the RAM is too small. Should I increase the RAM for the test or disable the test?

garydgregory commented:
https://issues.apache.org/jira/browse/IO-764

garydgregory reviewed (changes requested):
-1 for now due to test failures. See comments. Also rebase on master to pick up a different version of the test: `org.apache.commons.io.IOUtilsTest.testBigString()`

michael-o commented on src/test/java/org/apache/commons/io/IOUtilsTest.java:
```diff
@@ -1643,5 +1644,19 @@ public void testToString_URL_CharsetName() throws Exception {
+    
+    @Test
+    public void testBigString() throws IOException {
+       String foo = StringUtils.repeat("\uD83D", (Integer.MAX_VALUE)/2-1);
```
No shared CI will provide you these resources.

garydgregory replied:
> No shared CI will provide you these resources.

Hi @michael-o 
I wrote a different version of the test that is less resource-intensive and more lenient. See git master. The main code is still broken though.

DaGeRe commented:
Thanks for the hints. I've removed my test, added the comments, removed closing of the channel and fixed the checkstyle problems, and I rebased to latest master state.

Getting the behaviour by mocking is not possible from my point of view, because the problem happens when `getBytes` produces an array that is too big for handling - returning this array from a mocked `getBytes` would result in the same ressource usage as having it inside a String.

Is this ok from your point of view?

garydgregory commented:
Hi @DaGeRe 
Thank you for the update. 
Something is wrong with this PR. It contains unrelated changes you did not author.

DaGeRe commented:
Thanks for the hint, these were included on rebasing.

To fix this, I branched freshly from master, made the change and committed via force-push. 

Is this ok from now from your point of view?

garydgregory commented:
Now it git master.
